Fix default iface flannel

StefanoSalsano opened the pull request:
in this commit I've fixed the setup of the flannel plugin
it needs that /etc/hosts in the minion includes the resolution of the host name, otherwise all minion nodes are seen as 10.0.2.15 in the master

Fixes #51

ferrarimarco commented on scripts/linux/cleanup-k8s-and-cni.sh:
```diff
@@ -0,0 +1,40 @@
+#!/bin/sh
```
This script seems unrelated to this PR. Can you please remove it and open another one? Thanks!

Don't worry about adding a new commit, I'll squash everything before merging :)

StefanoSalsano replied:
done

ferrarimarco commented on scripts/linux/bootstrap-kubernetes-minion.sh:
```diff
@@ -1,14 +1,15 @@
 #!/bin/sh
 
+SCRIPT_PATH="/vagrant/scripts/linux"
```
Let me think if we can get rid of this hardwired path...

StefanoSalsano replied:
ok, see my comment at the end

ferrarimarco commented on scripts/linux/bootstrap-kubernetes-minion.sh:
```diff
@@ -24,8 +25,11 @@ elif [ "$network_plugin_id" = 'calico' ]; then
 
 
+echo "Initializing Kubernetes minion to join: $master_address and token: $token"
+kubeadm join "$master_address":6443 --token "$token" --discovery-token-unsafe-skip-ca-verification
```
Missing newline at the end :)

StefanoSalsano replied:
fixed

ferrarimarco commented on scripts/linux/config-etc-hosts.sh:
```diff
@@ -0,0 +1,31 @@
+SCRIPT_PATH="/vagrant/scripts/linux"
+
+PLAYGROUND_NAME="$1"
+#PLAYGROUND_NAME="k8s-p9"
```
Remove this commented line

StefanoSalsano replied:
fixed

ferrarimarco commented on scripts/linux/config-etc-hosts.sh:
```diff
@@ -0,0 +1,31 @@
+
+
+SCRIPT_NAME="manage-hosts.sh"
+SCRIPT_PATH="/vagrant/scripts/linux"
```
See above. Let me think if we can get rid of this. Also, this is a duplicate...

StefanoSalsano replied:
ok, see my comment at the end

ferrarimarco commented on scripts/linux/config-etc-hosts.sh:
```diff
@@ -0,0 +1,31 @@
+MINION_2="k8s-minion-2"
+MINION_3="k8s-minion-3"
+
+MASTER_1_IP="192.168.0.10"
```
This is not going to work if you use a different subnet. Additionally, this is not scalable if we add/remove hosts (TBH, we have this issue elsewhere as well)

StefanoSalsano replied:
ok, see my comment at the end

ferrarimarco commented on scripts/linux/config-etc-hosts.sh:
```diff
@@ -0,0 +1,31 @@
+$SCRIPT_PATH/$SCRIPT_NAME remove $MINION_2.$PLAYGROUND_NAME.local
+$SCRIPT_PATH/$SCRIPT_NAME remove $MINION_3.$PLAYGROUND_NAME.local
+
+$SCRIPT_PATH/$SCRIPT_NAME add $MASTER_1.$PLAYGROUND_NAME.local $MASTER_1_IP
```
I get what you're doing here, but I guess it's easier to manage with Ansible...

See https://docs.ansible.com/ansible/latest/modules/lineinfile_module.html#examples

ferrarimarco replied:
I can work on this bit

StefanoSalsano replied:
ok, see my comment at the end

ferrarimarco commented on defaults.yaml:
```diff
@@ -20,6 +20,7 @@ net:
   minion_1_ipv6_part: "c41e::"
   minion_2_ipv6_part: "c41f::"
   minion_3_ipv6_part: "c420::"
+  if_name_for_flannel: enp0s8
```
The interface name is not stable :(

We have to get it dynamically.

StefanoSalsano replied:
ok, see my comment at the end

ferrarimarco commented:
Also, is this related to #52 ?

StefanoSalsano commented:
this PR has a set of shortcomings... I was aware of these shortcomings :-)
anyway these shortcomings only affect the flannel plugin, and it is important to have it working to start the work on performance modeling/measurements
I will identify the shortcomings hereafter, I suggest that we merge the pull request, copy the list of shortcomings into a new issue and keep working on solving the shortcomings in parallel with the performance modeling/measurements

1) bootstrap-kubernetes-minion.sh uses config-etc-hosts.sh to setup /etc/hosts file, but the names/ip addresses of hosts are statically configured in config-etc-hosts.sh
this can be made more dynamic... also it could be easier to manage with Ansible with a different approach
note that this is not scalable if we add/remove hosts (but this issue is already present)

2) defaults.yaml includes an hardwired reference to the "data plane" interface of nodes
if_name_for_flannel: enp0s8
we should get it dynamically 
@StefanoSalsano : maybe this setting is not even needed if the /etc/hosts file is correctly set

ferrarimarco commented:
After some debugging, I think we're doing something wrong.

My guess is that flannel is not picking up the iface setting.

In fact, the actual values of the hosts file don't actually matter. In this test I commented the `$SCRIPT_PATH/config-etc-hosts.sh "$playground_name"` in the minion bootstrap script. After that, I executed that command manually on the master and on worker 1.

```
[vagrant@k8s-master-1 ~]$ kubectl get nodes -o wide
NAME                          STATUS   ROLES    AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE                KERNEL-VERSION               CONTAINER-RUNTIME
k8s-master-1.k8s-play.local   Ready    master   10m   v1.17.3   10.0.0.10     <none>        CentOS Linux 7 (Core)   3.10.0-693.21.1.el7.x86_64   docker://19.3.7
k8s-minion-1.k8s-play.local   Ready    <none>   10m   v1.17.3   10.0.0.30     <none>        CentOS Linux 7 (Core)   3.10.0-693.21.1.el7.x86_64   docker://19.3.7
k8s-minion-2.k8s-play.local   Ready    <none>   10m   v1.17.3   10.0.2.15     <none>        CentOS Linux 7 (Core)   3.10.0-693.21.1.el7.x86_64   docker://19.3.7
k8s-minion-3.k8s-play.local   Ready    <none>   10m   v1.17.3   10.0.2.15     <none>        CentOS Linux 7 (Core)   3.10.0-693.21.1.el7.x86_64   docker://19.3.7
[vagrant@k8s-master-1 ~]$ cat /etc/hosts
127.0.0.1       base-box-builder.k8s-play.local base-box-builder
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
10.0.0.10 registry.k8s-play.local
        k8s-master-1.k8s-play.local 192.168.0.10
        k8s-minion-1.k8s-play.local 192.168.0.30
        k8s-minion-2.k8s-play.local 192.168.0.31
        k8s-minion-3.k8s-play.local 192.168.0.32
```

My env file:

```
net:
  network_prefix: "10.0.0."
  network_prefix_ipv6: "fde4:8dba:82e1:"
  subnet_mask: "255.255.255.0"
  subnet_mask_ipv6: "56"
  master_ipv6_part: "c40A::"
  minion_1_ipv6_part: "c41e::"
  minion_2_ipv6_part: "c41f::"
  minion_3_ipv6_part: "c420::"
  if_name_for_flannel: enp0s8
ansible:
  group_vars:
    all:
      kubernetes_network_plugin: flannel
    kubernetes-masters:
    kubernetes-minions:
```

ferrarimarco commented:
@StefanoSalsano have a look at my latest commits :)

Don't worry about the history, I'll squash everything as needed. Thanks!

StefanoSalsano commented on README.md:
```diff
@@ -25,8 +25,8 @@ This project is a playground to play with Kubernetes.
 1. Prepare a base Vagrant box.
-   1. Provision and configure a VM (`vagrant up base-box-builder.k8s-play.local`).
-   1. Export a Vagrant box based on the `vagrant up base-box-builder.k8s-play.local` VM.
+   1. Provision and configure a VM (`vagrant base-box-builder.k8s-play.local`).
```
are you sure that "up" is not needed? if I remember correctly I run:
vagrant up base-box-builder.k8s-play.local
but I have not tried now

ferrarimarco replied:
Hope now it's clearer (added one commit)

StefanoSalsano commented:
I've tested it... we're getting close but there is still some issue:
I have destroyed master and minions and I have run vagrant up (I have NOT rebuilt the basebox)

I have tested with flannel plugin

the output of kubectl get pod --namespace kube-system  -o wide
is NOT OK, because the pods on the master still have the 10.0.2.15 address (while the pods on the minions are ok)

NAME                                                READY   STATUS    RESTARTS   AGE     IP             NODE                        NOMINATED NODE   READINESS GATES
etcd-k8s-master-1.k8s-p9.local                      1/1     Running   0          9m45s   10.0.2.15      k8s-master-1.k8s-p9.local   <none>           <none>
kube-apiserver-k8s-master-1.k8s-p9.local            1/1     Running   0          9m45s   10.0.2.15      k8s-master-1.k8s-p9.local   <none>           <none>
kube-controller-manager-k8s-master-1.k8s-p9.local   1/1     Running   0          9m45s   10.0.2.15      k8s-master-1.k8s-p9.local   <none>           <none>
kube-flannel-ds-amd64-44blk                         1/1     Running   0          9m30s   10.0.2.15      k8s-master-1.k8s-p9.local   <none>           <none>
kube-flannel-ds-amd64-4l2vm                         1/1     Running   0          9m25s   192.168.0.32   k8s-minion-3.k8s-p9.local   <none>           <none>
kube-flannel-ds-amd64-8jfmk                         1/1     Running   0          9m25s   192.168.0.31   k8s-minion-2.k8s-p9.local   <none>           <none>
kube-flannel-ds-amd64-8tvfp                         1/1     Running   0          9m15s   192.168.0.30   k8s-minion-1.k8s-p9.local   <none>           <none>
kube-proxy-288ds                                    1/1     Running   0          9m25s   192.168.0.31   k8s-minion-2.k8s-p9.local   <none>           <none>
kube-proxy-b5l92                                    1/1     Running   0          9m15s   192.168.0.30   k8s-minion-1.k8s-p9.local   <none>           <none>
kube-proxy-kwbrq                                    1/1     Running   0          9m30s   10.0.2.15      k8s-master-1.k8s-p9.local   <none>           <none>
kube-proxy-s86dw                                    1/1     Running   0          9m25s   10.0.2.15      k8s-minion-3.k8s-p9.local   <none>           <none>
kube-scheduler-k8s-master-1.k8s-p9.local            1/1     Running   0          9m45s   10.0.2.15      k8s-master-1.k8s-p9.local   <none>           <none>

StefanoSalsano commented:
maybe I have understood... in kubernetes.yml the setup of FQDN is done as last step ? I'm trying now to move it earlier

StefanoSalsano commented:
BINGO!!

StefanoSalsano commented:
still something unexpected... the /etc/hosts on the master is 

127.0.0.1       k8s-master-1.k8s-p9.local       k8s-master-1
127.0.0.1       base-box-builder.k8s-p7.local   base-box-builder
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
192.168.0.10 registry.k8s-p9.local

the line 192.168.0.30 k8s-master-1.k8s-p9.local is missing

the output of ansible is different in the master and in the minions
TASK [Add the FQDN to the hosts file] ******************************************
task path: /etc/ansible/playbooks/kubernetes.yml:13
changed: [192.168.0.10] => {"backup": "", "changed": true, "msg": "line replaced"}
changed: [192.168.0.32] => {"backup": "", "changed": true, "msg": "line added"}
changed: [192.168.0.30] => {"backup": "", "changed": true, "msg": "line added"}
changed: [192.168.0.31] => {"backup": "", "changed": true, "msg": "line added"}

StefanoSalsano commented:
I think that the new problem is the interaction with the task "Configure registry name resolution in hosts file"

StefanoSalsano commented:
the ansible task blindly replaces a line with the IP address, so when the "Configure registry name resolution in hosts file" adds the registry, it deletes the entry for the master, because they have the same IP address

StefanoSalsano commented:
I'm trying with a solution based on the manage-hosts.sh script which has the logic to handle adding/removing entries from /etc/hosts

StefanoSalsano commented:
why do we need this entry ?
127.0.0.1 k8s-master-1.k8s-p9.local

I've found a solution by dropping this entry and replacing it with the correct one
192.168.0.10 k8s-master-1.k8s-p9.local

StefanoSalsano commented:
ho provato a deployare il daemonset iperf con flannel e va perfetto

ho provato anche con Calico con la nuova configurazione, parte tutto perfettamente, gli indirizzi 10.0.2.15 sono scomparsi, dopo che e' partito ho anche provato il daemonset iperf ed e' OK!

e' ora di andare a dormire prima che faccia giorno :-)

ferrarimarco commented:
Thanks Stefano for your efforts. I'd like to solve this using Ansible only, instead of relying on other shell scripts. Thanks for pointing me in the right direction tho!

ferrarimarco commented:
Desired hosts file end state:

```
[vagrant@k8s-master-1 ~]$ cat /etc/hosts
127.0.0.1       base-box-builder.k8s-play.local base-box-builder
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
10.0.0.10 registry.k8s-play.local
10.0.0.10       k8s-master-1.k8s-play.local
```

ferrarimarco commented:
@StefanoSalsano I was able to manage everything with Ansible, so I removed the unneeded scripts.

End state of the hosts file after the last commit:

```
[vagrant@k8s-master-1 ~]$ cat /etc/hosts
127.0.0.1       base-box-builder.k8s-play.local base-box-builder
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
10.0.0.10 registry.k8s-play.local
10.0.0.10 k8s-master-1.k8s-play.local
```

Merging now!
